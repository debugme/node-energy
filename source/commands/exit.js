const command = () => {
  process.exit(0)
}

export { command }
