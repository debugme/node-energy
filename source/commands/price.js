import roundTo from 'round-to'

import { isPositiveInteger } from '../utilities/utilities'
import { DECIMAL_PLACES, DAYS_IN_YEAR, PENCE_IN_POUND, MAX_THRESHOLD } from '../constants/constants'

const getTotalCost = (scratchpad, rateInfo) => {
  const { standing_charge = 0 } = scratchpad
  const { threshold = MAX_THRESHOLD, price } = rateInfo
  const delta = Math.min(scratchpad.balance, threshold)
  scratchpad.balance -= delta
  scratchpad.costInPence += (delta * price)
  scratchpad.costInPence += (DAYS_IN_YEAR * standing_charge)
  return scratchpad
}

const getPlanInfo = (scratchpad, planInfo) => {
  const { usage, vatMultiplier } = scratchpad
  const { supplier, plan, rates, standing_charge } = planInfo
  const options = { costInPence: 0, balance: usage, standing_charge, vatMultiplier }
  const { costInPence } = rates.reduce(getTotalCost, options)
  const costInPounds = costInPence * vatMultiplier / PENCE_IN_POUND
  const totalCost = roundTo(costInPounds, DECIMAL_PLACES)
  const row = { supplier, plan, totalCost }
  scratchpad.table.push(row)
  return scratchpad
}

const parseData = (options) => {
  const { tokens } = options
  if (tokens.length < 2)
    throw new Error('format is "price <annual_usage>"')
  const [, annualUsage] = tokens
  if (!isPositiveInteger(annualUsage))
    throw new Error(`annual usage "${annualUsage}" is not a positive integer amount`)
  return tokens
}

const command = (options) => {
  const { context: { plans, vatMultiplier } } = options
  const [, annualUsage] = parseData(options)
  const usage = parseInt(annualUsage, 10)
  const scratchpad = plans.reduce(getPlanInfo, { usage, vatMultiplier, table: [] })
  scratchpad.table.sort((left, right) => left.totalCost - right.totalCost)
  return scratchpad.table
}

export { command }
