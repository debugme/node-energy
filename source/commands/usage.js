import { isPositiveInteger, findByMatch, findByQuery } from '../utilities/utilities'
import { MONTHS_IN_YEAR, DAYS_IN_YEAR, PENCE_IN_POUND, MAX_THRESHOLD } from '../constants/constants'

const updateUsage = (scratchpad, rate) => {
  const { vatMultiplier } = scratchpad
  const { price, threshold = MAX_THRESHOLD } = rate
  const costPerUnitOfEnergy = (price * vatMultiplier)
  for (let index = 0; index < threshold; ++index) {
    if (scratchpad.balance - costPerUnitOfEnergy < 0)
      break
    scratchpad.balance -= costPerUnitOfEnergy
    scratchpad.usage++
  }
  return scratchpad
}

const getUsage = (values) => {
  const { monthlySpend, planInfo: { rates, standing_charge = 0 }, vatMultiplier } = values
  const annualCharge = standing_charge * DAYS_IN_YEAR * vatMultiplier
  const annualSpend = monthlySpend * MONTHS_IN_YEAR * PENCE_IN_POUND
  const usage = 0
  const balance = annualSpend - annualCharge
  const scratchpad = rates.reduce(updateUsage, { usage, balance, vatMultiplier })
  return scratchpad.usage
}

const parseData = (options) => {
  const { tokens, context: { plans } }  = options
  if (tokens.length < 4)
    throw new Error('format is "usage <supplier> <plan> <spend>"')
  const [ , supplier, plan, spend ] = tokens
  const findBySupplier = findByMatch.bind(this, 'supplier', supplier)
  const findByPlan = findByMatch.bind(this, 'plan', plan)
  const isValidSpend = isPositiveInteger(spend)
  if (!plans.find(findBySupplier))
    throw new Error(`supplier "${supplier}" not found in existing plans`)
  if (!plans.find(findByPlan))
    throw new Error(`plan "${plan}" not found in existing plans`)
  if (!isValidSpend)
    throw new Error(`spend "${spend}" is not a positive integer amount`)
  return tokens
}

const command = (options) => {
  const { context: { plans, vatMultiplier } }  = options
  const [ , supplier, plan, spend ] = parseData(options)
  const query = { supplier, plan }
  const findByPlan = findByQuery.bind(this, query)
  const monthlySpend = parseInt(spend, 10)
  const planInfo = plans.find(findByPlan)
  if (!planInfo)
    throw new Error(`plan matching query "${JSON.stringify(query)}" not found in existing plans`)
  const values = { vatMultiplier, monthlySpend, planInfo }
  return getUsage(values)
}

export { command }
