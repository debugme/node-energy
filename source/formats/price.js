const toString = ({supplier, plan, totalCost}) =>
  `${supplier},${plan},${totalCost}`

const format = (table) =>
  table.map(toString).join('\n')

export { format }