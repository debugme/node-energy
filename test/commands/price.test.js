import { command } from '../../source/commands/price'
import plans from '../../data/plans'

describe('the price command function', () => {

  let options = null

  beforeEach(() => {
    options = { tokens: [/* each test to fill in */], context: { plans, 'vatMultiplier': 1.05 }}
  })

  it('should generate a table when given valid inputs', () => {
    options.tokens = ['price', '1000']
    expect(command(options)).toMatchSnapshot()
  })

  it('should throw an error if wrong number of arguments were passed in', () => {
    options.tokens = ['price']
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toMatchSnapshot()
  })

  it('should throw an error if invalid annual usage was passed in', () => {
    options.tokens = ['price', '10.5']
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toMatchSnapshot()
  })

})
