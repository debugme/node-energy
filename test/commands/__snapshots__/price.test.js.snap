// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the price command function should generate a table when given valid inputs 1`] = `
Array [
  Object {
    "plan": "variable",
    "supplier": "eon",
    "totalCost": 108.68,
  },
  Object {
    "plan": "fixed",
    "supplier": "edf",
    "totalCost": 111.25,
  },
  Object {
    "plan": "standard",
    "supplier": "ovo",
    "totalCost": 120.23,
  },
  Object {
    "plan": "standing-charge",
    "supplier": "bg",
    "totalCost": 121.33,
  },
]
`;

exports[`the price command function should throw an error if invalid annual usage was passed in 1`] = `"Error: annual usage \\"10.5\\" is not a positive integer amount"`;

exports[`the price command function should throw an error if wrong number of arguments were passed in 1`] = `"Error: format is \\"price <annual_usage>\\""`;
